src/Analyzer/Passes/AggregateFunctionOfGroupByKeysPass.cpp

```diff
@@ -0,0 +1,191 @@
+    {
+        return parent->as<ListNode>();
+    }
+
+    void visitFunction(const FunctionNode * function_node)
+    {
+        if (function_node->getArguments().getNodes().empty())
+        {
+            keep_aggregator = true;
+            return;
+        }
+        auto it = std::find_if(
+            group_by_keys.begin(),
+            group_by_keys.end(),
+            [function_node](const QueryTreeNodePtr & group_by_ele) { return function_node->isEqual(*group_by_ele); });
+
+        if (it == group_by_keys.end())
+        {
+            KeepEliminateFunctionVisitor visitor(group_by_keys, keep_aggregator);
+            visitor.visit(function_node->getArgumentsNode());
+        }
+    }
+
+    void visitColumn(const ColumnNode * column)
+    {
+        /// if variable of a function is not in GROUP BY keys, this function should not be deleted
+        auto it = std::find_if(
+            group_by_keys.begin(),
+            group_by_keys.end(),
+            [column](const QueryTreeNodePtr & group_by_ele) { return column->isEqual(*group_by_ele); });
+
+        if (it == group_by_keys.end())
+            keep_aggregator = true;
+    }
+
+    void visitImpl(const QueryTreeNodePtr & node)
+    {
+        if (keep_aggregator)
+            return;
+
+        if (node->as<ListNode>())
+            return;
+
+        if (auto * function_node = node->as<FunctionNode>())
+        {
+            visitFunction(function_node);
+        }
+        else if (auto * column = node->as<ColumnNode>())
+        {
+            visitColumn(column);
+        }
+    }
+
+private :
+    const QueryTreeNodes & group_by_keys;
+    bool & keep_aggregator;
+};
+
+/// Try to eliminate min/max/any/anyLast which will not decent into subqueries.
+class EliminateFunctionVisitor : public InDepthQueryTreeVisitor<EliminateFunctionVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitor<EliminateFunctionVisitor>;
+    using Base::Base;
+
+    EliminateFunctionVisitor(const QueryTreeNodes & group_by_keys_) : group_by_keys(group_by_keys_) { }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+
+        /// Check if function is min/max/any/anyLast
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node
+            || !(function_node->getFunctionName() == "min" || function_node->getFunctionName() == "max"
+                 || function_node->getFunctionName() == "any" || function_node->getFunctionName() == "anyLast"))
+            return;
+
+        if (!function_node->getArguments().getNodes().empty())
+        {
+            bool keep_aggregator = false;
+
+            KeepEliminateFunctionVisitor visitor(group_by_keys, keep_aggregator);
+            visitor.visit(function_node->getArgumentsNode());
+
+            /// Place argument of an aggregate function instead of function
+            if (!keep_aggregator)
+                node = function_node->getArguments().getNodes()[0];
```

Comment:
Also, it can be broken because of `group_by_use_nulls`

Reply:
hi @novikd, I tested `group_by_use_nulls` and it seems ok, if there is something wrong, could you explain more?

```
SELECT a, any(b) FROM (SELECT number % 2 as a, number % 3 as b FROM numbers(1)) GROUP BY GROUPING SETS ((a, b), (a)) settings group_by_use_nulls=1;

Query id: 1a2230ad-98ad-4ea0-8d65-e3128f996573

┌─a─┬─any(b)─┐
│ 0 │   ᴺᵁᴸᴸ │
└───┴────────┘
┌─a─┬─any(b)─┐
│ 0 │      0 │
└───┴────────┘

explain query tree SELECT a, any(b) FROM (SELECT number % 2 as a, number % 3 as b FROM numbers(1)) GROUP BY GROUPING SETS ((a, b), (a)) settings group_by_use_nulls=1;

┌─explain───────────────────────────────────────────────────────────────────────────────────────┐
│ QUERY id: 0, group_by_type: grouping_sets                                                     │
│   PROJECTION COLUMNS                                                                          │
│     a Nullable(UInt8)                                                                         │
│     any(b) UInt8                                                                              │
│   PROJECTION                                                                                  │
│     LIST id: 1, nodes: 2                                                                      │
│       COLUMN id: 2, column_name: a, result_type: Nullable(UInt8), source_id: 3                │
│       COLUMN id: 4, column_name: b, result_type: UInt8, source_id: 3                          │
│   JOIN TREE                                                                                   │
│     QUERY id: 3, is_subquery: 1                                                               │
│       PROJECTION COLUMNS                                                                      │
......
```